packages/@aws-cdk/aws-iam/lib/private/precreated-role.ts

```diff
@@ -0,0 +1,305 @@
+    this.roleArn = this.role.roleArn;
+    this.roleName = this.role.roleName;
+    this.stack = this.role.stack;
+
+    Dependable.implement(this, {
+      dependencyRoots: [this.role],
+    });
+
+    // add a single PolicySynthesizer under the `App` scope
+    this.policySynthesizer = (this.node.root.node.tryFindChild(POLICY_SYNTHESIZER_ID)
+      ?? new PolicySynthesizer(this.node.root)) as PolicySynthesizer;
+    this.policySynthesizer.addRole(this.node.path, {
+      roleName: props.missing ? 'missing role' : this.roleName,
+      managedPolicies: this.managedPolicies,
+      policyStatements: this.policyStatements,
+      assumeRolePolicy: Stack.of(this).resolve(props.assumeRolePolicy?.toJSON()?.Statement),
+    });
+  }
+
+  public attachInlinePolicy(policy: Policy): void {
+    const statements = policy.document.toJSON()?.Statement;
+    if (statements && Array.isArray(statements)) {
+      statements.forEach(statement => {
+        this.policyStatements.push(this.resolveJsonObject(statement));
+      });
+    }
+  }
+
+  public addManagedPolicy(policy: IManagedPolicy): void {
+    this.managedPolicies.push(this.resolveReferences(policy.managedPolicyArn));
+  }
+
+  public addToPolicy(statement: PolicyStatement): boolean {
+    this.policyStatements.push(this.resolveReferences(statement.toStatementJson()));
+    return false;
+  }
+
+  public addToPrincipalPolicy(statement: PolicyStatement): AddToPrincipalPolicyResult {
+    this.addToPolicy(statement);
+    // If we return `false`, the grants will try to add the statement to the resource
+    // (if possible).
+    return { statementAdded: true, policyDependable: new DependencyGroup() };
+  }
+
+  public grant(grantee: IPrincipal, ...actions: string[]): Grant {
+    return this.role.grant(grantee, ...actions);
+  }
+
+  public grantPassRole(grantee: IPrincipal): Grant {
+    return this.role.grantPassRole(grantee);
+  }
+
+  public grantAssumeRole(identity: IPrincipal): Grant {
+    return this.role.grantAssumeRole(identity);
+  }
+
+  /**
+   * Resolve any references and replace with a more user friendly value. This is the value
+   * that will appear in the report, so instead of getting something like this (not very useful):
+   *
+   *     "Resource": {
+   *       "Fn::Join": [
+   *         "",
+   *         [
+   *           "arn:",
+   *           {
+   *             "Ref": "AWS::Partition"
+   *           },
+   *           ":iam::",
+   *           {
+   *             "Ref": "AWS::AccountId"
+   *           },
+   *           ":role/Role"
+   *         ]
+   *       ]
+   *     }
+   *
+   * We will instead get:
+   *
+   *     "Resource": "arn:(PARTITION):iam::(ACCOUNT):role/Role"
+   *
+   * Or if referencing a resource attribute
+   *
+   *     "Resource": {
+   *       "Fn::GetAtt": [
+   *         "SomeResource",
+   *         "Arn"
+   *       ]
+   *     }
+   *
+   * Becomes
+   *
+   *     "(Path/To/SomeResource.Arn)"
```

Comment:
Note: wrt the debate of the LogicalID that's generated via CFn, as a recommended best practice vs the 2-step deploy, grab the ARNs, fill them in the policy.

Scenario: deploy the roles, grab the ARNs for the resources, fill in the policies with the actual values, add them to the prod stack to the role for explicit and complete boundary.

Q1: A deployment that changes the Logical ids, will it also invalidate the policies? Would they have to be re-edited?
Q2: Can the teams front-load some of the work by having a list of resources with fixed logicalIds and know that they will have to 2-step deploy if names change? Does the list of resources make it more transparent which policy will be rendered invalid too (if the assumption is correct)?

Reply:
* Q1: yes. If logical IDs change, resources must be replaced which requires a name change (hence invalidating the policies)
* Q2: Not sure I understand the Q.

I think the nature of this method brings with it that you're going to have to wildcard your resources. We're not changing that. The only thing we're doing is adding more information that you otherwise would have had.

Reply:
Q2 was an after thought for an edge case, let me add context:

If there is a customer with a security team enforcing the condition not to wildcard anything, then both roles and policies' content could be done with one change if the logical ids (names) are known beforehand.

Probably the question would be: is there a case that would justify hardcoding names in order to unblock work? 

Thinking about it, the answer becomes more and more `no`. As any issue with such a setup would get a swift reference to the page containing best practices and recommendations :-)